Revert commit 
 ef9caeccae "MIPS: Fix race condition with FPU thread task flag during context switch."

---
 arch/mips/kernel/r2300_switch.S |   15 ++++++++++++---
 arch/mips/kernel/r4k_switch.S   |   12 +++++++++---
 include/asm-mips/system.h       |   12 +++---------
 3 files changed, 24 insertions(+), 15 deletions(-)

diff --git a/arch/mips/kernel/r2300_switch.S b/arch/mips/kernel/r2300_switch.S
--- a/arch/mips/kernel/r2300_switch.S
+++ b/arch/mips/kernel/r2300_switch.S
@@ -43,7 +43,7 @@
 
 /*
  * task_struct *resume(task_struct *prev, task_struct *next,
- *                     struct thread_info *next_ti, int usedfpu)
+ *                     struct thread_info *next_ti) )
  */
 LEAF(resume)
 #ifndef CONFIG_CPU_HAS_LLSC
@@ -54,9 +54,18 @@ LEAF(resume)
 	cpu_save_nonscratch a0
 	sw	ra, THREAD_REG31(a0)
 
-	beqz	a3, 1f
+	/*
+	 * check if we need to save FPU registers
+	 */
+	lw	t3, TASK_THREAD_INFO(a0)
+	lw	t0, TI_FLAGS(t3)
+	li	t1, _TIF_USEDFPU
+	and	t2, t0, t1
+	beqz	t2, 1f
+	nor	t1, zero, t1
 
-	PTR_L	t3, TASK_THREAD_INFO(a0)
+	and	t0, t0, t1
+	sw	t0, TI_FLAGS(t3)
 
 	/*
 	 * clear saved user stack CU1 bit
diff --git a/arch/mips/kernel/r4k_switch.S b/arch/mips/kernel/r4k_switch.S
--- a/arch/mips/kernel/r4k_switch.S
+++ b/arch/mips/kernel/r4k_switch.S
@@ -41,7 +41,7 @@
 
 /*
  * task_struct *resume(task_struct *prev, task_struct *next,
- *                     struct thread_info *next_ti, int usedfpu)
+ *                     struct thread_info *next_ti)
  */
 	.align	5
 	LEAF(resume)
@@ -56,10 +56,16 @@
 	/*
 	 * check if we need to save FPU registers
 	 */
+	PTR_L	t3, TASK_THREAD_INFO(a0)
+	LONG_L	t0, TI_FLAGS(t3)
+	li	t1, _TIF_USEDFPU
+	and	t2, t0, t1
+	beqz	t2, 1f
+	nor	t1, zero, t1
 
-	beqz    a3, 1f
+	and	t0, t0, t1
+	LONG_S	t0, TI_FLAGS(t3)
 
-	PTR_L	t3, TASK_THREAD_INFO(a0)
 	/*
 	 * clear saved user stack CU1 bit
 	 */
diff --git a/include/asm-mips/system.h b/include/asm-mips/system.h
--- a/include/asm-mips/system.h
+++ b/include/asm-mips/system.h
@@ -28,7 +28,7 @@
  * switch_to(n) should switch tasks to task nr n, first
  * checking that n isn't the current task, in which case it does nothing.
  */
-extern asmlinkage void *resume(void *last, void *next, void *next_ti, u32 __usedfpu);
+extern asmlinkage void *resume(void *last, void *next, void *next_ti);
 
 struct task_struct;
 
@@ -48,8 +48,6 @@ struct task_struct;
 
 #define switch_to(prev,next,last)					\
 do {									\
-	u32 __usedfpu;							\
-									\
 	if (cpu_has_fpu &&						\
 	    (prev->thread.mflags & MF_FPUBOUND) &&			\
 	     (!(KSTK_STATUS(prev) & ST0_CU1))) {			\
@@ -59,19 +57,15 @@ do {									\
 	if (cpu_has_dsp)						\
 		__save_dsp(prev);					\
 	next->thread.emulated_fp = 0;					\
-	__usedfpu = test_and_clear_tsk_thread_flag(prev, TIF_USEDFPU);	\
-	(last) = resume(prev, next, task_thread_info(next), __usedfpu);	\
+	(last) = resume(prev, next, task_thread_info(next));		\
 } while(0)
 
 #else
 #define switch_to(prev,next,last)					\
 do {									\
-	u32 __usedfpu;							\
-									\
 	if (cpu_has_dsp)						\
 		__save_dsp(prev);					\
-	__usedfpu = test_and_clear_tsk_thread_flag(prev, TIF_USEDFPU);	\
-	(last) = resume(prev, next, task_thread_info(next), __usedfpu);	\
+	(last) = resume(prev, next, task_thread_info(next));		\
 } while (0)
 #endif
 
-- 
